### ISSUE TITLE
Collecting PyTorch -> Flux migration notes

### ISSUE THREAD

**Issue by BioTurboNick** [FluxML/Flux.jl#2410]
I'm in the process of moving a model from PyTorch to Flux, and I'm going to catalog the challenges I've found in migrating, perhaps for a doc page or to improve docs generally. If others wish to add anything, please do!

Weights initialization:
In PyTorch, the default weight init method is `kaiming_uniform` aka He initialization.
In Flux, the default weight init method is `glorot_uniform` aka Xavier initialization.`

PyTorch chooses a `gain` for the init function based on the type of nonlinearity specified, which defaults to `leaky_relu`, and uses this:
```
a = √5 # argument passed into PyTorch's `kaiming_uniform_` function, the "negative slope" of the rectifier
gain = √(2 / (1 + a ^ 2))
```
Flux defaults the `kaiming_uniform` gain to `√2`, which is what PyTorch would use if `relu` was specified rather than `leaky_relu`.

To replicate the PyTorch default, `kaiming_uniform(gain = √(2 / (1 + a ^ 2)))` can be provided for the `init` keyword argument of the layer constructor.

Layers:
In PyTorch, there are separate objects for different dimensionality (e.g. `conv1d1, conv2d`, `conv3d`. In Flux, the dimensionality is specified by the tuple provided for the kernel.

In PyTorch, a convolution layer will automatically pick a stride, so that e.g. a 2d convolution with a 2x2 kernel will halve the frame dimensions. With Flux, the stride must be specified explicitly; here, stride = 2.

In PyTorch, activation functions are inserted as separate steps in the chain, as equals to layers. In Flux, they are provided as an argument to the layer constructor.

`Sequential` => `Chain`

`Linear` => `Dense`

When building a custom layer, the `(::MyLayer)(input) = ` method is the equivalent of `def forward(self, input):`

Often if PyTorch has a method, Flux (many via MLUtils.jl) has the same method. e.g. `unsqueeze` to insert a 1-length dimension, or `erf` to compute the error function. Note that the inverse of `unsqueeze` is actually `Base.dropdims`.

Training:
A single step in Flux is simply `gradient` followed by `update!`. In PyTorch there are more steps: the optimizer's gradients must be zeroed with `.zero_grad()`, then the loss is calculated, then the tensor returned from the loss function is backward-propagated with `.backward()`, and finally the optimizer is stepped forward with `.step()`. In Flux, both parts can be combined with the `train!` function, and can also be used to iterate over a set of paired training inputs and outputs.

In Flux, an optimizer state object is first obtained by `setup`, and this state is passed to the training loop. In PyTorch, the optimizer object itself is manipulated in the training loop.

**Comment by darsnack** [comment]
@BioTurboNick Reported on Slack that with the default Flux initialization, his model would get stuck in an all zeros state, but not with the PyTorch init.

**Comment by CarloLucibello** [comment]
Regarding pytorch's initializations, there is a long discussion 
https://github.com/pytorch/pytorch/issues/18182
on how to update their default initialization to current best practice. 

- There seems to be consensus that zero bias is a good init, so we should stick to that.
- Pytorch initizializes with kaiming uniform [-1/sqrt(in_features), 1/sqrt(in_features)], while Flux's does glorot uniform [-sqrt(6 / (in_features + out_features)), +sqrt(6/(in_features + out_features))]. I couldn't find any convincing paper supporting the choice of one over the other.

I would therefore keep initialization as it is.

**Comment by BioTurboNick** [comment]
Just added a few more differences. Thanks to Claude Opus 5 for allowing me to systematically compare and contrast a Flux (attempt at) an implementation of an existing PyTorch model bit by bit. Was driving me crazy for years that it still didn't match PyTorch and we had to go back to using that component in Python.

**Comment by CarloLucibello** [comment]
Nice. We should add a documentation page with all that
